### PR TITLE
feat(localization): add `pose_instability_detector`

### PR DESCRIPTION
**Pull request by SakodaShintaro**
## Description

Added `pose_instability_detector` node.

See [README.md](https://github.com/SakodaShintaro/autoware.universe/tree/feat/add_pose_instability_detector/localization/pose_instability_detector) for details.

A known problem is that the amount of data published to `/diagnostics` in the current autoware is too large and can cause it to drop some data.

## Tests performed

The following tests have been run and verified

* run autoware on sample rosbag
* run autoware on AWSIM's rosbag and the error between ekf and GT is the same as before.
* colcon test

```
# if use lcov, need following libraries
# sudo apt-get install -y lcov python3-colcon-lcov-result

colcon build --symlink-install \
  --cmake-args \
    -DCMAKE_BUILD_TYPE=Release \
    -DBUILD_TESTING=ON \
    -DCMAKE_CXX_FLAGS='-fprofile-arcs -ftest-coverage' \
    -DCMAKE_C_FLAGS='-fprofile-arcs -ftest-coverage' \
  --packages-select pose_instability_detector
colcon test --return-code-on-test-failure --packages-select pose_instability_detector
colcon lcov-result --packages-select pose_instability_detector
colcon test-result --all --verbose | grep pose_instability_detector
```

A directory named `lcov` will be created under `~/autoware`, where coverage can be checked.

## Result
The following results are obtained when run in AWSIM's rosbag.

diff_pose

## Effects on system behavior

An output to check the stability of the EKF output values will be added to the diagnostics.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: localization/pose_instability_detector/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+| Name                  | Type   | Description                         |
+| --------------------- | ------ | ----------------------------------- |
+| `interval_sec`        | double | Interval of timer callback (second) |
+| `threshold_linear_x`  | double | Threshold of linear x (m)           |
+| `threshold_linear_y`  | double | Threshold of linear y (m)           |
+| `threshold_linear_z`  | double | Threshold of linear z (m)           |
+| `threshold_angular_x` | double | Threshold of angular x (rad)        |
+| `threshold_angular_y` | double | Threshold of angular y (rad)        |
+| `threshold_angular_z` | double | Threshold of angular z (rad)        |
```
Comment (Motsu-san) [inline localization/pose_instability_detector/README.md:31]: I think it is better following [this guide using json schema](https://github.com/orgs/autowarefoundation/discussions/3371).
How do you think?

**Review by Motsu-san — CHANGES_REQUESTED** [review]
Thank you for your pull-request!
I left some comment. Please check them.

File: localization/pose_instability_detector/src/pose_instability_detector.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    std::bind(&PoseInstabilityDetector::callback_odometry, this, std::placeholders::_1));
+
+  twist_sub_ = this->create_subscription<TwistWithCovarianceStamped>(
+    "~/input/twist", 10,
+    std::bind(&PoseInstabilityDetector::callback_twist, this, std::placeholders::_1));
+
+  const double interval_sec = this->declare_parameter<double>("interval_sec");
+  timer_ = this->create_wall_timer(
+    std::chrono::duration<double>(interval_sec),
+    std::bind(&PoseInstabilityDetector::callback_timer, this));
```
Comment (KYabuuchi) [inline localization/pose_instability_detector/src/pose_instability_detector.cpp:46]: Is it intentional to use a wall_timer? 
Using a wall_timer might lead to unexpected intervals for verification, especially when playing rosbags or testing in AWSIM at a lower speed.

FYI: The following code is how to create a callback that follows `/clock` when `use_sim_time` is true and *wall_time* when false.
```cpp
  timer_ = rclcpp::create_timer(
    this, this->get_clock(), std::chrono::duration<double>(interval_sec),
    std::bind(&PoseInstabilityDetector::callback_timer, this));
```

File: localization/pose_instability_detector/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+This node triggers periodic timer callbacks to compare two poses:
+
+- The pose obtained by integrating the twist values from the last received message on `/localization/kinematic_state` over a duration specified by `interval_sec`.
+- The latest pose from `/localization/kinematic_state`.
+
+The results of this comparison are then output to the `/diagnostics` topic.
+
+If this node outputs WARN messages to `/diagnostics`, it means that the EKF output is significantly different from the integrated twist values.
+This discrepancy suggests that there may be an issue with either the `pose_estimator` or the `twist_estimator`.
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply localization/pose_instability_detector/README.md]: Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5439/commits/52559b5d0306c736673e6dac98c7c4aef8f921f4

File: localization/pose_instability_detector/config/pose_instability_detector.param.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/**:
+  ros__parameters:
+    interval_sec: 1.0  # [sec]
+    threshold_linear_x: 0.1  # [m]
+    threshold_linear_y: 0.1  # [m]
+    threshold_linear_z: 0.1  # [m]
+    threshold_angular_x: 0.1  # [rad]
+    threshold_angular_y: 0.1  # [rad]
+    threshold_angular_z: 0.1  # [rad]
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply localization/pose_instability_detector/config/pose_instability_detector.param.yaml]: Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5439/commits/e06c66add3c1ccb6c46c7902032a24033fe5d686

File: localization/pose_instability_detector/README.md (hunk 1 above)
Comment (SakodaShintaro) [reply localization/pose_instability_detector/README.md]: Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5439/commits/e06c66add3c1ccb6c46c7902032a24033fe5d686.

`mkdocs serve` in `autoware/src/universe/autoware.universe` and go to <http://127.0.0.1:8000/autoware.universe/localization/pose_instability_detector/>

I can see a following table.

Screenshot from 2023-10-31 13-51-42


and the schema check is passed
https://github.com/autowarefoundation/autoware.universe/actions/runs/6702224534/job/18210896634?pr=5439

File: localization/pose_instability_detector/src/pose_instability_detector.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  // compare pose and latest_odometry_
+  const Pose latest_ekf_pose = latest_odometry_.pose.pose;
+  PoseStamped ekf_to_odom;
+  ekf_to_odom.pose = tier4_autoware_utils::inverseTransformPose(pose, latest_ekf_pose);
+  ekf_to_odom.header.stamp = latest_odometry_.header.stamp;
+  ekf_to_odom.header.frame_id = "base_link";
+  const geometry_msgs::msg::Point pos = ekf_to_odom.pose.position;
+  const auto [ang_x, ang_y, ang_z] = quat_to_rpy(ekf_to_odom.pose.orientation);
```
Comment (YamatoAndo) [inline localization/pose_instability_detector/src/pose_instability_detector.cpp:135]: It seems like the header information is not necessary, so I believe you can write simpler.
```suggestion
  Pose ekf_to_odom;
  ekf_to_odom = tier4_autoware_utils::inverseTransformPose(pose, latest_ekf_pose);
  const geometry_msgs::msg::Point pos = ekf_to_odom.position;
  const auto [ang_x, ang_y, ang_z] = quat_to_rpy(ekf_to_odom.orientation);
```

File: localization/pose_instability_detector/src/pose_instability_detector.cpp [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+  const std::vector<std::string> labels = {"diff_position_x", "diff_position_y", "diff_position_z",
+                                           "diff_angle_x",    "diff_angle_y",    "diff_angle_z"};
+
+  // publish diagnostics
+  DiagnosticArray diagnostics;
+  diagnostics.header.stamp = stamp;
+
+  DiagnosticStatus status;
+  status.name = "pose_instability_detector";
```
Comment (YamatoAndo) [inline localization/pose_instability_detector/src/pose_instability_detector.cpp:159]: I'd like to include "localization: " as a prefix to make the name of the diagnostic.
```suggestion
  status.name = "localization: pose_instability_detector";
```

Comment (YamatoAndo) [inline localization/pose_instability_detector/src/pose_instability_detector.cpp:159]: Please add hardware_id.
I'm considering putting the node name in the hardware_id.

`status.hardware_id = "pose_instability_detector";`
or 
`status.hardware_id = this->get_name();`

**Comment by YamatoAndo** [comment]
@SakodaShintaro Please update the rqt_runtime_monitor.png to its latest state just before merging.

File: localization/pose_instability_detector/src/pose_instability_detector.cpp (hunk 5 above)
Comment (SakodaShintaro) [reply localization/pose_instability_detector/src/pose_instability_detector.cpp]: That is correct. Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5439/commits/4b2ebe79c5fe063ea9f79a00b61c9157978a0612

File: localization/pose_instability_detector/src/pose_instability_detector.cpp (hunk 6 above)
Comment (SakodaShintaro) [reply localization/pose_instability_detector/src/pose_instability_detector.cpp]: Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5439/commits/62ffb708f2291af2d550980e702e7599eda95b01

**Review by Motsu-san — APPROVED** [review]
[LGTMeow](https://lgtmeow.com)

**Comment by SakodaShintaro** [comment]
@KYabuuchi @YamatoAndo 

Thank you for comments.
Have all the problems been resolved?

It seems that there is a change to launch, so it cannot be merged without the approve of the person who has the authority to do so.
The algorithm of the code has already been reviewed by Sakamoto-san.

**Review by KYabuuchi — APPROVED** [review]

**Pull request merged**